webknossos-datastore/app/com/scalableminds/webknossos/datastore/models/datasource/DataLayer.scala

```diff
@@ -227,6 +228,51 @@ trait DataLayerLike {
   // n-dimensional datasets = 3-dimensional datasets with additional coordinate axes
   def additionalAxes: Option[Seq[AdditionalAxis]]
 
+  // Datasets that are not in the WKW format use mags
+  def magsOpt: Option[List[MagLocator]] = this match {
+    case layer: AbstractDataLayer         => layer.mags
+    case layer: AbstractSegmentationLayer => layer.mags
+    case layer: DataLayerWithMagLocators  => Some(layer.getMags)
+    case _                                => None
+  }
+
+  def dataFormatOpt: Option[DataFormat.Value] = this match {
+    case layer: WKWDataLayer                 => Some(layer.dataFormat)
+    case layer: WKWSegmentationLayer         => Some(layer.dataFormat)
+    case layer: ZarrDataLayer                => Some(layer.dataFormat)
+    case layer: ZarrSegmentationLayer        => Some(layer.dataFormat)
+    case layer: N5DataLayer                  => Some(layer.dataFormat)
+    case layer: N5SegmentationLayer          => Some(layer.dataFormat)
+    case layer: PrecomputedDataLayer         => Some(layer.dataFormat)
+    case layer: PrecomputedSegmentationLayer => Some(layer.dataFormat)
+    case layer: Zarr3DataLayer               => Some(layer.dataFormat)
+    case layer: Zarr3SegmentationLayer       => Some(layer.dataFormat)
+    // Abstract layers
+    case _ => None
+  }
+
+  def numChannelsOpt: Option[Int] = this match {
+    case layer: AbstractDataLayer            => layer.numChannels
+    case layer: AbstractSegmentationLayer    => layer.numChannels
+    case layer: ZarrDataLayer                => layer.numChannels
+    case layer: ZarrSegmentationLayer        => layer.numChannels
+    case layer: N5DataLayer                  => layer.numChannels
+    case layer: N5SegmentationLayer          => layer.numChannels
+    case layer: PrecomputedDataLayer         => layer.numChannels
+    case layer: PrecomputedSegmentationLayer => layer.numChannels
+    case layer: Zarr3DataLayer               => layer.numChannels
+    case layer: Zarr3SegmentationLayer       => layer.numChannels
+    case _                                   => None
+  }
+
+  def wkwResolutionsOpt: Option[List[WKWResolution]] = this match {
+    case layer: AbstractDataLayer         => layer.wkwResolutions
+    case layer: AbstractSegmentationLayer => layer.wkwResolutions
+    case layer: WKWDataLayer              => Some(layer.wkwResolutions)
+    case layer: WKWSegmentationLayer      => Some(layer.wkwResolutions)
+    case _                                => None
+  }
```

Comment:
The woes of case classes… Unfortunately I don’t know how to compact this further.

Reply:
This could be compacted further by having a common trait which can be matched to. But not sure whether introducing another trait to the whole data layer hierarchy would be ideal :thinking:

Reply:
side note: That would be one advantage of scala 3 (still not voting for migrating to scala 3 though)